File: built-in plugins dsp/mb-util-tuner.cpp
/**
 * MB Chromatic Tuner
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Precision chromatic tuner with reference pitch adjustment
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_UTIL_TUNER_H
#define MB_UTIL_TUNER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbUtilTuner : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-util-tuner";
    static constexpr const char* PLUGIN_NAME    = "MB Chromatic Tuner";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float reference = 440f;  // range [420, 460]
    float tolerance = 5f;  // range [1, 20]
    float mute = 0f;  // range [0, 1]
    };

    MbUtilTuner() = default;
    ~MbUtilTuner() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.reference = std::clamp(params.reference, 420f, 460f);
        params.tolerance = std::clamp(params.tolerance, 1f, 20f);
        params.mute = std::clamp(params.mute, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Chromatic Tuner
        return input;
    }
};

#endif // MB_UTIL_TUNER_H

File: built-in plugins dsp/mb-util-tuner.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbUtilTunerPlugin: PluginDefinition = {
  id: "mb-util-tuner",
  slug: "mb-util-tuner",
  name: "MB Chromatic Tuner",
  category: "effect",
  type: "eq" as any,
  version: "1.0.0",
  description: "Precision chromatic tuner with reference pitch adjustment",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "reference",
      name: "Reference Pitch",
      type: "float",
      defaultValue: 440,
      minValue: 420,
      maxValue: 460,
      automatable: false,
    },
    {
      id: "tolerance",
      name: "Tolerance",
      type: "float",
      defaultValue: 5,
      minValue: 1,
      maxValue: 20,
      automatable: false,
    },
    {
      id: "mute",
      name: "Mute Through",
      type: "float",
      defaultValue: 0,
      minValue: 0,
      maxValue: 1,
      automatable: false,
    },
  ],
  defaultPreset: { reference: 440, tolerance: 5, mute: 0 },
};

export default MbUtilTunerPlugin;
